src/libm-tester/testerutil.c

```diff
@@ -0,0 +1,230 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <mpfr.h>
+#include <time.h>
+#include <float.h>
+#include <limits.h>
+#include <math.h>
+
+#define DENORMAL_DBL_MIN (4.9406564584124654418e-324)
+#define POSITIVE_INFINITY INFINITY
+#define NEGATIVE_INFINITY (-INFINITY)
+
+#define DENORMAL_FLT_MIN (1.4012984643248170709e-45f)
+#define POSITIVE_INFINITYf ((float)INFINITY)
+#define NEGATIVE_INFINITYf (-(float)INFINITY)
+
+int isnumber(double x) { return !isinf(x) && !isnan(x); }
+int isPlusZero(double x) { return x == 0 && copysign(1, x) == 1; }
+int isMinusZero(double x) { return x == 0 && copysign(1, x) == -1; }
+double sign(double d) { return d < 0 ? -1 : 1; }
+int xisnan(double x) { return x != x; }
+
+int isnumberf(float x) { return !isinff(x) && !isnanf(x); }
+int isPlusZerof(float x) { return x == 0 && copysignf(1, x) == 1; }
+int isMinusZerof(float x) { return x == 0 && copysignf(1, x) == -1; }
+float signf(float d) { return d < 0 ? -1 : 1; }
+int xisnanf(float x) { return x != x; }
+
+int enableFlushToZero = 0;
+
+double flushToZero(double y) {
+  if (enableFlushToZero && fabs(y) < FLT_MIN) y = copysign(0.0, y);
+  return y;
+}
+
+int cmpDenormsp(float x, mpfr_t fry) {
+  float y = mpfr_get_d(fry, GMP_RNDN);
+  x = flushToZero(x);
+  y = flushToZero(y);
+  if (xisnanf(x) && xisnanf(y)) return 1;
+  if (xisnanf(x) || xisnanf(y)) return 0;
+  if (isinf(x) != isinf(y)) return 0;
+  if (x == POSITIVE_INFINITYf && y == POSITIVE_INFINITYf) return 1;
+  if (x == NEGATIVE_INFINITYf && y == NEGATIVE_INFINITYf) return 1;
+  if (y == 0) {
+    if (isPlusZerof(x) && isPlusZerof(y)) return 1;
+    if (isMinusZerof(x) && isMinusZerof(y)) return 1;
+    return 0;
+  }
+  if (!xisnanf(x) && !xisnanf(y) && !isinf(x) && !isinf(y)) return signf(x) == signf(y);
+  return 0;
+}
+
+int cmpDenormdp(double x, mpfr_t fry) {
+  double y = mpfr_get_d(fry, GMP_RNDN);
+  if (xisnan(x) && xisnan(y)) return 1;
+  if (xisnan(x) || xisnan(y)) return 0;
+  if (isinf(x) != isinf(y)) return 0;
+  if (x == POSITIVE_INFINITY && y == POSITIVE_INFINITY) return 1;
+  if (x == NEGATIVE_INFINITY && y == NEGATIVE_INFINITY) return 1;
+  if (y == 0) {
+    if (isPlusZero(x) && isPlusZero(y)) return 1;
+    if (isMinusZero(x) && isMinusZero(y)) return 1;
+    return 0;
+  }
+  if (!xisnan(x) && !xisnan(y) && !isinf(x) && !isinf(y)) return sign(x) == sign(y);
+  return 0;
+}
+
+int firstTime = 1;
+mpfr_t fra, frb, frc, frd;
+
+double countULPdp(double d, mpfr_t c) {
```

Comment:
countULP2dp and countULPdp are different. countULP2dp does not report error above DBL_MIN. This is sometimes needed since some functions do not guarantee error bound below DBL_MIN. What do you mean by merging those two functions?